### PR TITLE
Add x509 test framework and happy-path rountrip write tests

### PR DESCRIPTION
**Pull request by bbrks**
### Prerequisite PRs:
- [x] #4576 

### Description:
- Adds X.509 test helpers to generate and provision server/client certs in tests
- Adds 2 happy-path roundtrip write tests (write a doc, wait for DCP) for following scenarios:
  - IP address only connection/cert
  - Domain only connection/cert

### Test Config:
The following test flags have been added for the X.509 tests:
- `SG_TEST_X509` - Default: false - When true, X.509 tests are run
- `SG_TEST_X509_SSH_USERNAME` - Default: root - Used by SSH to copy cert files to Couchbase Server. 
  - User must have write access to /opt/couchbase/...
  - User must have a valid SSH Key configured with the server

### CODE REVIEW

File: rest/x509_utils_test.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+	cmd = exec.Command("ssh", forceKeyOnly, skipHostFingerprint, sshRemoteHost, "chmod -R a+rx", filepath.Join(destinationDirectory, filepath.Base(sourceFilepath)))
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return errors.Wrap(err, string(output))
+	}
+
+	return nil
+}
+
+// domainForIPAddr returns a domain name that publicly resolves to the given IP address.
+func domainForIPAddr(ip net.IP) string {
```
Comment (adamcfraser) [inline rest/x509_utils_test.go]: This is a nice trick!  Does it add any requirements for the test environment that we need to worry about (e.g. public internet access)?

**Review by adamcfraser — DISMISSED** [review]
Looks good!  A few questions/clarifications/nit picks, but feel free to merge if you don't think changes are merited.

File: rest/x509_utils_test.go (hunk 1 above)
Comment (bbrks) [reply rest/x509_utils_test.go]: Oops, this was used when I was automatically renaming the cluster node, but not required any more... Still using xip.io for testing this, but manually instead.

It does require public internet access to resolve them. For completely isolated test environments I'd imagine we can use /etc/hosts or IT's internal DNS to provide local-only test domains

File: rest/x509_utils_test.go [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+	if err := enableX509ClientCertsInCouchbaseServer(restAPIURL); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func couchbaseNodeConfiguredHostname(restAPIURL url.URL) (string, error) {
+	// Configure CBS to enable optional X.509 client certs
```
Comment (bbrks) [inline rest/x509_utils_test.go]:
```suggestion
```

File: rest/x509_utils_test.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	base.Debugf(base.KeyAll, "triggered reload of certificates on Couchbase Server")
+
+	if err := enableX509ClientCertsInCouchbaseServer(restAPIURL); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func couchbaseNodeConfiguredHostname(restAPIURL url.URL) (string, error) {
```
Comment (bbrks) [inline rest/x509_utils_test.go:293]:
```suggestion
// couchbaseNodeConfiguredHostname returns the Couchbase node name for the given URL.
func couchbaseNodeConfiguredHostname(restAPIURL url.URL) (string, error) {
```

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**